app/xinit/startx.man
.\" $Xorg: startx.man,v 1.4 2001/02/09 02:05:49 xorgcvs Exp $
.\" $XdotOrg$
.\" Copyright 1993, 1998  The Open Group
.\" 
.\" Permission to use, copy, modify, distribute, and sell this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" Except as contained in this notice, the name of The Open Group shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from The Open Group.
.\"
.\" $XFree86: xc/programs/xinit/startx.man,v 1.7 2001/04/19 15:08:32 dawes Exp $
.\"
.TH STARTX 1 __xorgversion__
.SH NAME
startx \- initialize an X session
.SH SYNOPSIS
.B startx
[ [
.I client
]
.I options
\&\.\|.\|. ] [
.B \-\^\-
[
.I server
]
.I options
\&.\|.\|. ]
.SH DESCRIPTION
The \fIstartx\fP script is a front end to \fIxinit\fP that provides a
somewhat nicer user interface for running a single session of the X
Window System.  It is often run with no arguments.
.PP
Arguments immediately following the
.I startx
command are used to start a client in the same manner as
.IR xinit (1).
The special argument
.RB '--'
marks the end of client arguments and the beginning of server options.
It may be convenient to specify server options with startx to change on a
per-session basis the
default color depth, the server's notion of the number of dots-per-inch the
display device presents, or take advantage of a different server layout, as
permitted by the
.IR __XSERVERNAME__ (1)
server and specified in the
.IR __XCONFIGFILE__ (__filemansuffix__)
file.  Some examples of specifying server arguments follow; consult the
manual page for your X server to determine which arguments are legal.
.RS
.PP
startx -- -depth 16
.PP
startx -- -dpi 100
.PP
startx -- -layout Multihead
.RE
#ifdef __SCOMAN__
.PP
To determine the client to run,
.I startx
looks for the following files, in order:
.RS
.PP
.I $(HOME)/.startxrc
.PP
.I __libdir__/sys.startxrc
.PP
.I $(HOME)/.xinitrc
.PP
.I __xinitdir__/xinitrc
.RE
.PP
#else
.PP
To determine the client to run,
.I startx
first looks for a file called
.I .xinitrc
in the user's home directory.  If that is not found, it uses
the file
.I xinitrc
in the
.I xinit
library directory.
#endif
If command line client options are given, they override this
behavior and revert to the
.IR xinit (1)
behavior.
To determine the server to run,
.I startx
first looks for a file called
.I .xserverrc
in the user's home directory.  If that is not found, it uses
the file
.I xserverrc
in the
.I xinit
library directory.
If command line server options are given, they override this
behavior and revert to the
.IR xinit (1)
behavior.  Users rarely need to provide a
.I .xserverrc
file.
See the
.IR xinit (1)
manual page for more details on the arguments.
.PP
The system-wide
.I xinitrc
and
.I xserverrc
files are found in the
.I __xinitdir__
directory.
.PP
The
.I .xinitrc
is typically a shell script which starts many clients according to the
user's preference.  When this shell script exits,
.I startx
kills the server and performs any other session shutdown needed.
Most of the clients started by
.I .xinitrc
should be run in the background.  The last client should run in the
foreground; when it exits, the session will exit.  People often choose
a session manager, window manager, or \fIxterm\fP as the ''magic'' client.
.SH EXAMPLE
.PP
Below is a sample \fI\.xinitrc\fP that starts several applications and
leaves the window manager running as the ''last'' application.  Assuming that
the window manager has been configured properly, the user
then chooses the ''Exit'' menu item to shut down X.
.sp
.in +4
.nf
xrdb \-load $HOME/.Xresources
xsetroot \-solid gray &
xbiff \-geometry \-430+5 &
oclock \-geometry 75x75\-0\-0 &
xload \-geometry \-80\-0 &
xterm \-geometry +0+60 \-ls &
xterm \-geometry +0\-100 &
xconsole \-geometry \-0+0 \-fn 5x7 &
exec twm
.fi
.in -4
.SH "ENVIRONMENT VARIABLES"
.TP 25
DISPLAY
This variable gets set to the name of the display to which clients should
connect.  Note that this gets
.IR set ,
not read.
.TP 25
XAUTHORITY
This variable, if not already defined, gets set to
.IR $(HOME)/.Xauthority .
This is to prevent the X server, if not given the
.I \-auth
argument, from automatically setting up insecure host-based authentication
for the local host.  See the
.IR Xserver (1)
and
.IR Xsecurity (__miscmansuffix__)
manual pages for more information on X client/server authentication.
.SH FILES
#ifdef __SCOMAN__
.TP 25
.I $(HOME)/.startxrc
Client to run.  Typically a shell script which runs many programs in
the background.
.TP 25
.I __libdir__/sys.startxrc
Client to use if the user has no
.I .startxrc
file.
#endif
.TP 25
.I $(HOME)/.xinitrc
Client to run.  Typically a shell script which runs many programs in
the background.
.TP 25
.I $(HOME)/.xserverrc
Server to run.  The default is
.IR X .
.TP 25
.I __xinitdir__/xinitrc
Client to run if the user has no
.I .xinitrc
file.
.TP 25
.I __xinitdir__/xserverrc
Server to run if the user has no
.I .xserverrc
file.
.SH "SEE ALSO"
.IR xinit (1),
.IR Xserver (1),
.IR __XSERVERNAME__ (1)


xserver/hw/xfree86/utils/xorgcfg/startx.c
/*
 * Copyright (c) 2000 by Conectiva S.A. (http://www.conectiva.com)
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *  
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL
 * CONECTIVA LINUX BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF
 * OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 * 
 * Except as contained in this notice, the name of Conectiva Linux shall
 * not be used in advertising or otherwise to promote the sale, use or other
 * dealings in this Software without prior written authorization from
 * Conectiva Linux.
 *
 * Author: Paulo César Pereira de Andrade <pcpa@conectiva.com.br>
 *
 */

#include "config.h"
#include <sys/types.h>
#include <sys/wait.h>
#include <signal.h>

/*
 * Initialization
 */
static int xpid;
Display *DPY;

/*
 * Implementation
 */
Bool
startx(void)
{
    int timeout = 8;

    if (getenv("DISPLAY") != NULL)
	/* already running Xserver */
	return (False);

    if (XF86Config_path == NULL) {
	char *home, filename[PATH_MAX];
	char commandline[PATH_MAX * 4];
	int c_pos;
	int len;
	/* 
	 * The name of the 4.0 binary is XFree86. X might also
	 * be the name of the 3.3 binary. Therefore don't change
	 * name to 'X'.
	 */
	if (XFree86_path)
	    c_pos = XmuSnprintf(commandline, sizeof(commandline),
				"%s/"__XSERVERNAME__" :8 -configure ",XFree86_path);
	else
	    c_pos = XmuSnprintf(commandline, sizeof(commandline), 
				"%s/bin/"__XSERVERNAME__" :8 -configure ", XFree86Dir);
	if (XF86Module_path && ((len = sizeof(commandline) - c_pos) > 0))
	    c_pos += XmuSnprintf(commandline + c_pos,len,
				 " -modulepath %s",XF86Module_path);
	if (XF86Font_path && ((len = sizeof(commandline) - c_pos) > 0))
	    c_pos += XmuSnprintf(commandline + c_pos,len,
				 " -fontpath %s",XF86Font_path);
	if (XF86RGB_path && ((len = sizeof(commandline) - c_pos) > 0))
	    c_pos += XmuSnprintf(commandline + c_pos,len,
				 " -rgbpath %s",XF86RGB_path);
	
	if (system(commandline) != 0) {
	    fprintf(stderr, "Failed to run \"X -configure\".\n");
	    exit(1);
	}

	if ((home = getenv("HOME")) == NULL)
	    home = "/";

#ifndef QNX4
	XmuSnprintf(filename, sizeof(filename), "%s/"__XCONFIGFILE__".new", home);
#else
	XmuSnprintf(filename, sizeof(filename), "//%d%s/"__XCONFIGFILE__".new",
		    getnid(), home);
#endif

	/* this memory is never released, even if the value of XF86Config_path is
	 * changed.
	 */
	XF86Config_path = XtNewString(filename);
    }

    putenv("DISPLAY=:8");

    switch (xpid = fork()) {
	case 0: {
	    char path[PATH_MAX];
	    /* Don't change to X! see above */
	    if (XFree86_path)
	        XmuSnprintf(path, sizeof(path), "%s/"__XSERVERNAME__, XFree86_path);
	    else
	        XmuSnprintf(path, sizeof(path), "%s/bin/"__XSERVERNAME__, XFree86Dir);
#ifndef X_PRIVSEP
	    execl(path, "X", ":8", /*"+xinerama",*/ "+accessx",
		"-allowMouseOpenFail", "-nolisten", "tcp",
		  "-xf86config", XF86Config_path, (void *)NULL);
#else
	    execl(path, "X", ":8", /*"+xinerama",*/ "+accessx",
		"-allowMouseOpenFail", "-keepPriv", "-nolisten", "tcp",
		  "-xf86config", XF86Config_path, (void *)NULL);
#endif
	    exit(-127);
	}   break;
	case -1:
	    fprintf(stderr, "Cannot fork.\n");
	    exit(1);
	    break;
	default:
	    break;
    }

    while (timeout > 0) {
	int status;

	sleep(timeout -= 2);
	if (waitpid(xpid, &status, WNOHANG | WUNTRACED) == xpid)
	    break;
	else {
	    DPY = XOpenDisplay(NULL);
	    if (DPY != NULL)
		break;
	}
    }

    if (DPY == NULL) {
	fprintf(stderr, "Cannot connect to X server.\n");
	exit(1);
    }

    return (True);
}

void
endx(void)
{
    if (xpid != 0)
	kill(xpid, SIGTERM);
}
